File: GoBackend/internal/controllers/map_controller.go
package controllers

import (

	"uber/internal/services"

	"github.com/gin-gonic/gin"
)

// 📍 COORDINATES
func GetCoordinates(c *gin.Context) {

	address := c.Query("address")
	if len(address) < 3 {
		c.JSON(400, gin.H{"message": "Invalid address"})
		return
	}

	data, err := services.GetAddressCoordinates(address)
	if err != nil {
		c.JSON(404, gin.H{"message": "Coordinates not found"})
		return
	}

	c.JSON(200, data)
}

// 🚗 DISTANCE
func GetDistanceTime(c *gin.Context) {

	origin := c.Query("origin")
	destination := c.Query("destination")

	if origin == "" || destination == "" {
		c.JSON(400, gin.H{"message": "Origin & destination required"})
		return
	}

	data, err := services.GetDistanceAndTime(origin, destination)
	if err != nil {
		c.JSON(500, gin.H{"message": err.Error()})
		return
	}

	c.JSON(200, data)
}

// 🔍 AUTOCOMPLETE
func GetAutoCompleteSuggestions(c *gin.Context) {

	input := c.Query("input")
	if input == "" {
		c.JSON(400, gin.H{"message": "Input required"})
		return
	}

	data, err := services.GetAutoCompleteSuggestions(input)
	if err != nil {
		c.JSON(500, gin.H{"message": "Failed to fetch suggestions"})
		return
	}

	c.JSON(200, data)
}

File: SpringServer/src/main/java/com/uber/springserver/controllers/MapController.java
package com.uber.springserver.controllers;

import com.uber.springserver.services.MapService;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/maps")
public class MapController {

    private final MapService mapService;

    public MapController(MapService mapService) {
        this.mapService = mapService;
    }

    @GetMapping("/get-coordinates")
    public Object getCoordinates(@RequestParam String address) {
        if (address == null || address.length() < 3) {
            return Map.of("message", "Invalid address");
        }
        try {
            return mapService.getAddressCoordinates(address);
        } catch (Exception e) {
            return Map.of("message", "Coordinates not found");
        }
    }

    @GetMapping("/get-distance-time")
    public Object getDistanceTime(@RequestParam String origin, @RequestParam String destination) {
        if (origin == null || origin.isBlank() || destination == null || destination.isBlank()) {
            return Map.of("message", "Origin & destination required");
        }
        try {
            return mapService.getDistanceAndTime(origin, destination);
        } catch (Exception e) {
            return Map.of("message", e.getMessage());
        }
    }

    @GetMapping("/get-suggestions")
    public Object getSuggestions(@RequestParam String input) {
        if (input == null || input.isBlank()) {
            return Map.of("message", "Input required");
        }
        try {
            List<Object> data = mapService.getAutoCompleteSuggestions(input);
            return data;
        } catch (Exception e) {
            return Map.of("message", "Failed to fetch suggestions");
        }
    }
}
